addon/components/ui-dropdown.js

```diff
@@ -1,40 +1,233 @@
 import Ember from 'ember';
 import Base from '../mixins/base';
-import DataAttributes from '../mixins/data-attributes';
+import PromiseResolver from '../mixins/promise-resolver';
 
-export default Ember.Component.extend(Base, DataAttributes, {
+const _proxyCallback = function(callbackName) {
+  return function(value, text, $element) {
+    return this.get(`attrs.${callbackName}`)(this._getObjectOrValue(value), text, $element, this);
+  };
+};
+
+export default Ember.Component.extend(Base, PromiseResolver, {
   module: 'dropdown',
-  classNames: [ 'ui', 'dropdown' ],
+  classNames: ['ui', 'dropdown'],
+  ignorableAttrs: ['selected'],
+  objectMap: null,
+
+  init() {
+    this._super(...arguments);
+    this.set('objectMap', {});
+  },
+
+  willDestroyElement() {
+    this._super(...arguments);
+    this.set('objectMap', null);
+  },
+
+  // Semantic Hooks
+  willInitSemantic(settings) {
+    this._super(...arguments);
+    if (settings.onChange) {
+      settings.onChange = this.get('_onChange');
+    }
+    if (settings.onAdd) {
+      settings.onAdd = this.get('_onAdd');
+    }
+    if (settings.onRemove) {
+      settings.onRemove = this.get('_onRemove');
+    }
+  },
+
+  didInitSemantic() {
+    this._super(...arguments);
+    // We want to handle this outside of the standard process
+    this.get('_settableAttrs').removeObject('selected');
+    // We need to ensure the internal value is set to '',
+    // otherwise when we get the value later it is undefined
+    // and semantic returns the module instead of the actual value
+    this.execute('clear');
+    this._inspectSelected();
+  },
 
-  didInsertElement() {
+  didUpdateAttrs() {
     this._super(...arguments);
-    var selected = this.get('selected');
-    if (Ember.isPresent(selected)) {
-      this.execute('set selected', this.getSelected(selected));
+    this._inspectSelected();
+  },
+
+  actions: {
+    mapping(object) {
+      let guid = Ember.guidFor(object);
+      if (!this.get('objectMap').hasOwnProperty(guid)) {
+        this.get('objectMap')[guid] = object;
+      }
+      Ember.run.scheduleOnce('afterRender', this, this._inspectSelected);
+      return guid;
+    }
+  },
+
+  // Method proxies
+  _onChange(value, text, $element) {
+    // Semantic calls the events on any 'set {action}'
+    // Because of that we want to ignore calls when we are
+    // Specifically setting the value
+    if (this.get('_isSettingSelect')) {
+      return;
+    }
+    let returnValue;
+    if (this.execute('is multiple')) {
+      let values = this.execute('get values');
+      returnValue = [];
+      for (let i = 0; i < Ember.get(values, 'length'); i++) {
+        let item = this._atIndex(values, i);
+        returnValue.push(this._getObjectOrValue(item));
+      }
+    } else {
+      returnValue = this._getObjectOrValue(value);
+    }
+
+    return this.attrs.onChange(returnValue, text, $element, this);
```

Comment:
Mostly I didn't since we already called the method _getObjectOrValue an above and I didn't want it processing again. In the case of the is_multiple, if we returned an empty array, the proxy method would return null, which wouldn't be the expected behavior. The onAdd and OnRemove always have a single call which helped use a standard method.